examples/interpreter-demo/backwards.html

```diff
@@ -0,0 +1,320 @@
+        </block>
+      </next>
+    </block>
+  </xml>
+
+  <script>
+    var demoWorkspace = Blockly.inject('blocklyDiv',
+        {media: 'https://unpkg.com/blockly/media/',
+         toolbox: document.getElementById('toolbox')});
+    Blockly.JavaScript.STATEMENT_PREFIX = 'highlightBlock(%1);\n';
+    Blockly.JavaScript.addReservedWords('highlightBlock');
+    Blockly.Xml.domToWorkspace(document.getElementById('startBlocks'),
+                               demoWorkspace);
+
+    var outputArea = document.getElementById('output');
+    var myInterpreter = null;
+
+    // Stack of serializations.  Each object contains three properties:
+    //   .json: A full serialization of the interpreter's state.
+    //   .delta: Instead of a full serialization, a delta from the next state.
+    //   .highlight: The ID of the block highlighted in this state.
+    var serializationStack = [];
+
+    // Global variable to keep track of the currently highlighted block.
+    var highlightedBlockId = '';
+
+    // The forward step button disables for 2s after program completion, then
+    // re-enables.  If the backwards step button is pressed during this time
+    // then cancel the scheduled re-enabling.
+    var disablePid = 0;
+
+    // Optionally, use the Diff Match Patch library to compress the stack.
+    var dmp;
+    if (typeof diff_match_patch === 'function') {
+      dmp = new diff_match_patch();
+      console.log('Using DMP for compression.');
+    } else {
+      console.warn('DMP not available, each step will cost ~300 KB.');
+    }
+
+    function initApi(interpreter, globalObject) {
+      // Add an API function for the alert() block, generated for "text_print" blocks.
+      var wrapper = function alert(text) {
+        text = arguments.length ? text : '';
+        outputArea.value += '\n' + text;
+      };
+      interpreter.setProperty(globalObject, 'alert',
+          interpreter.createNativeFunction(wrapper));
+
+      // Add an API function for the prompt() block.
+      var wrapper = function prompt(text) {
+        return window.prompt(text);
+      };
+      interpreter.setProperty(globalObject, 'prompt',
+          interpreter.createNativeFunction(wrapper));
+
+      // Add an API function for highlighting blocks.
+      var wrapper = function(id) {
+        id = String(id || '');
+        highlightedBlockId = id;
+        return highlightBlock(id);
+      };
+      interpreter.setProperty(globalObject, 'highlightBlock',
+          interpreter.createNativeFunction(wrapper));
+    }
+
+    var highlightPause = false;
+
+    function highlightBlock(id) {
+      demoWorkspace.highlightBlock(id);
+      highlightPause = true;
+    }
+
+    function resetStepUi(clearOutput) {
+      demoWorkspace.highlightBlock(null);
+      highlightPause = false;
+
+      if (clearOutput) {
+        outputArea.value = 'Program output:\n=================';
+      }
+      myInterpreter = null;
+    }
+
+    function stepBackwards() {
```

Comment:
> doesn't help developers who want to implement backwards debugging on their apps.

I see your point about how each app is going to need to store different data, however I disagree that option 2 it doesn't help. I can't think of a situation where an app (especially an educational one) *wouldn't* want to store data about the output of the previous step. Because if the user doesn't see the output of the previous step, it doesn't really help them with debugging, and it doesn't "feel" like the program has been stepped backwards. I think that giving some indication of how this would be done is important for external developers who want to implement the backwards stepping.

I'd also like to note that if the demo doesn't return to the previous output state, people may be confused about whether it is actually working as intended or not.

I feel like it should be possible to separate all of the output storage into separate functions so that it exists, gives the demo the "expected" functionality, and doesn't add an excessive amount of complexity. However, I didn't write the demo, so I'm happy for you to disagree on this.